contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -24,69 +27,131 @@ contract UmbraBatchSend {
     bytes32 ciphertext;
   }
 
+  mapping(address => uint256) internal totalTransferAmountPerToken;
+  struct TransferSummary {
+    uint256 amount;
+    address tokenAddr;
+  }
+
   error ValueMismatch();
+  error TransferAmmountMismatch();
   event BatchSendExecuted(address indexed sender);
 
   constructor(IUmbra _umbra) {
     umbra = _umbra;
   }
 
   function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) external payable {
-    if(msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
+    if (msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
     _batchSendEth(_tollCommitment, _params);
     emit BatchSendExecuted(msg.sender);
   }
 
-  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) external payable {
-    if(msg.value != _tollCommitment * _params.length) revert ValueMismatch();
-    _batchSendTokens(_tollCommitment, _params);
+  function batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    if (msg.value < _tollCommitment * _params.length) revert ValueMismatch();
+
+    for (uint256 i = 0; i < _params.length; i++) {
```

Comment:
Can use unchecked when incrementing here. If you want to dedupe a bit in all of the for loops you can do:

```solidity
function uncheckedIncrement(uint256 i) internal pure returns (uint256) {
  unchecked { return i + 1; }
}

for (uint256 i = 0; i < _params.length; i = uncheckedIncrement(i)) {
  // --- snip ---
}
```

Reply:
Looks like there's a loop in `batchSendTokens` that needs to be changed

Reply:
good catch, fixed